cmd/juju/machine/add.go

```diff
@@ -170,6 +171,44 @@ type MachineManagerAPI interface {
 
 var manualProvisioner = manual.ProvisionMachine
 
+// splitUserHost given a host string of example user@192.168.122.122
+// it will return user and 192.168.122.122
+func splitUserHost(host string) (string, string) {
+	if at := strings.Index(host, "@"); at != -1 {
+		return host[:at], host[at+1:]
+	}
+	return "", host
+}
+
+func (c *addCommand) scopeRun(client AddMachineAPI, config *config.Config, ctx *cmd.Context) error {
+	logger.Infof("manual provisioning")
+
+	authKeys, err := common.ReadAuthorizedKeys(ctx, "")
+	if err != nil {
+		return errors.Annotate(err, "reading authorized-keys")
+	}
+
+	user, host := splitUserHost(c.Placement.Directive)
+	args := manualcommon.ProvisionMachineArgs{
```

Comment:
@axw thanks for taking time from you holiday to enlighten me, I really appreciate. You have touched some valid points in this review but I don't think naming sshprovisioner/, winrmprovisioner/ is good, because in a couple of months windows will support ssh and linux like systems can in the future have the option to make powershell as the default shell. The ssh code already is very bashy syntax specific.

Anyway I will try to modify the current code to match your specification about the common interface and the dependency around packages.